File: backend/src/main/java/com/smartcampus/backend/module/resource/service/ResourceService.java
package com.smartcampus.backend.module.resource.service;

import com.smartcampus.backend.module.resource.dto.CreateResourceRequest;
import com.smartcampus.backend.module.resource.dto.ResourceResponse;
import com.smartcampus.backend.module.resource.dto.UpdateResourceRequest;
import com.smartcampus.backend.common.PageResponse;
import org.springframework.data.domain.Pageable;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Map;

public interface ResourceService {

    ResourceResponse         createResource(CreateResourceRequest request);
    ResourceResponse         getResourceById(String id);
    PageResponse<ResourceResponse> getAllResources(Map<String, String> filters, Pageable pageable);
    ResourceResponse         updateResource(String id, UpdateResourceRequest request);
    ResourceResponse         updateStatus(String id, String status);
    void                     deleteResource(String id);
    List<Map<String, Object>> checkAvailability(String id, LocalDateTime from, LocalDateTime to);
    List<String>             getResourceTypes();
}

File: frontend/src/features/resources/services/resourceService.js
import apiClient from '../../../services/apiClient';

const unwrapResponse = (responseData) => {
  if (!responseData) return responseData;
  if (Array.isArray(responseData)) return responseData;

  // Supports PageResponse directly: { content, totalElements, ... }
  if (responseData.content && Array.isArray(responseData.content)) {
    return responseData;
  }

  // Supports wrapped form: { data: ... }
  if (responseData.data) {
    return unwrapResponse(responseData.data);
  }

  return responseData;
};

export const resourceService = {
  getAllResources: async (filters = {}) => {
    const { page = 0, size = 10, type, status, location, capacity, search } = filters;
    const params = new URLSearchParams();

    if (type) params.append('type', type);
    if (status) params.append('status', status);
    if (location) params.append('location', location);
    if (capacity) params.append('capacity', capacity);
    if (search) params.append('search', search);
    params.append('page', String(page));
    params.append('size', String(size));

    const response = await apiClient.get(`/api/resources?${params.toString()}`);
    return unwrapResponse(response.data);
  },

  getResourceById: async (id) => {
    const response = await apiClient.get(`/api/resources/${id}`);
    return unwrapResponse(response.data);
  },

  createResource: async (resourceData) => {
    const response = await apiClient.post('/api/resources', resourceData);
    return unwrapResponse(response.data);
  },

  updateResource: async (id, resourceData) => {
    const response = await apiClient.put(`/api/resources/${id}`, resourceData);
    return unwrapResponse(response.data);
  },

  updateResourceStatus: async (id, status) => {
    const response = await apiClient.patch(`/api/resources/${id}/status`, { status });
    return unwrapResponse(response.data);
  },

  deleteResource: async (id) => {
    await apiClient.delete(`/api/resources/${id}`);
  },

  checkAvailability: async (id, from, to) => {
    const params = new URLSearchParams();
    if (from) params.append('from', from);
    if (to) params.append('to', to);
    const response = await apiClient.get(`/api/resources/${id}/availability?${params.toString()}`);
    return unwrapResponse(response.data);
  },

  getResourceTypes: async () => {
    const response = await apiClient.get('/api/resources/types');
    return unwrapResponse(response.data);
  },
  
  uploadResourceFile: async (file, folder = 'resources') => {
    const formData = new FormData();
    formData.append('file', file);
    formData.append('folder', folder);

    const response = await apiClient.post('/api/resources/upload', formData, {
      headers: {
        'Content-Type': undefined, // Tell Axios to let the browser set the boundary
      },
    });
    return unwrapResponse(response.data);
  },
};
